cpp/oneapi/dal/table/backend/csr_kernels.cpp

```diff
@@ -0,0 +1,163 @@
+
+    // overflows checked here
+    check_origin_data(origin_data, origin_info.element_count_, origin_dtype_size, block_dtype_size);
+
+    const std::int64_t origin_offset =
+        origin_row_indices[block_info.row_offset_] - origin_row_indices[0];
+
+    const std::int64_t block_size =
+        origin_row_indices[block_info.row_offset_ + block_info.row_count_] -
+        origin_row_indices[block_info.row_offset_];
+
+    const bool same_type(block_dtype == origin_info.dtype_);
+
+    // data
+    if (same_type && !alloc_kind_requires_copy(get_alloc_kind(origin_data), kind)) {
+        refer_origin_data(origin_data,
+                          origin_offset * block_dtype_size,
+                          block_size,
+                          block.data,
+                          preserve_mutability);
+    }
+    else {
+        if (block.data.get_count() < block_size || !block.data.has_mutable_data() ||
+            alloc_kind_requires_copy(get_alloc_kind(block.data), kind)) {
+            reset_array(policy, block.data, block_size, kind);
+        }
+
+        auto src_data = origin_data.get_data() + origin_offset * origin_dtype_size;
+        auto dst_data = block.data.get_mutable_data();
+
+        backend::convert_vector(policy,
+                                src_data + origin_offset * block_dtype_size,
+                                dst_data,
+                                origin_info.dtype_,
+                                block_dtype,
+                                block_size);
+    }
+
+    // column indices
+    if (!alloc_kind_requires_copy(get_alloc_kind(origin_column_indices), kind)) {
+        reset_array(policy, block.column_indices, block_size, kind);
+    }
+    refer_origin_data(origin_column_indices,
+                      origin_offset,
+                      block_size,
+                      block.column_indices,
+                      preserve_mutability);
+
+    // row_indices
+    if (block.row_indices.get_count() < block_info.row_count_ + 1 ||
+        !block.row_indices.has_mutable_data() ||
+        alloc_kind_requires_copy(get_alloc_kind(block.row_indices), kind)) {
+        reset_array(policy, block.row_indices, block_info.row_count_ + 1, kind);
+    }
+    if (block_info.row_offset_ == 0) {
+        refer_origin_data(origin_row_indices,
+                          0,
+                          block_info.row_count_,
+                          block.row_indices,
+                          preserve_mutability);
+    }
+    else {
+        auto src_row_indices = origin_row_indices.get_data();
+        auto dst_row_indices = block.row_indices.get_mutable_data();
+
+        for (std::int64_t i = 0; i < block_info.row_count_ + 1; i++) {
+            dst_row_indices[i] = src_row_indices[block_info.row_offset_ + i] -
+                                 src_row_indices[block_info.row_offset_] + 1;
+        }
+        block.row_indices.reset(origin_row_indices, dst_row_indices, block_info.row_count_ + 1);
+    }
```

Comment:
Not clear why the logic of pulling indices is so different. We can discuss it offline

Reply:
I'll try to explain. 1. Pull data is required compare of types and cast the types if needed (for indices it isn't required), the data isn't copy, addition memory isn't allocated. 2. For pulling column indices we don't have to compare of types. 3. For pulling row indices we have to allocate addition memory due to it's impossible to refer of base row indices. We can discuss it offline, yes